src/mpi/init/init_impl.c

```diff
@@ -107,17 +109,118 @@ const char **MPIR_pset_list = default_pset_list;
+
+/**
+ * @brief   Set the requested thread level of a session based on an info object
+ *
+ *          If the provided info_ptr is a nullptr or the thread_level key is not
+ *          found in the info object, the value of threadlevel remains unchanged.
+ *
+ * @param   info_ptr        Info object pointer to be parsed for thread_level key
+ * @param   threadlevel     Thread level found in info object (output)
+ *
+ * @return  int             MPI_SUCCESS or error code.
+ */
+static
+int get_thread_level_from_info(MPIR_Info * info_ptr, int *threadlevel)
+{
+    int mpi_errno = MPI_SUCCESS;
+    int buflen = 0;
+    char *thread_level_s = NULL;
+    int flag = 0;
+    const char key[] = "thread_level";
+
+    /* No info pointer, nothing todo here */
+    if (info_ptr == NULL) {
+        goto fn_exit;
+    }
+
+    /* Get the length of the thread level value */
+    mpi_errno = MPIR_Info_get_valuelen_impl(info_ptr, key, &buflen, &flag);
+    MPIR_ERR_CHECK(mpi_errno);
+
+    if (!flag) {
+        /* Key thread_level not found in info object */
+        goto fn_exit;
+    }
+
+    /* Get thread level value. No need to check flag afterwards
+     * because we would not be here if thread_level key was not in info object.
+     */
+    thread_level_s = MPL_malloc(buflen + 1, MPL_MEM_BUFFER);
+    mpi_errno = MPIR_Info_get_impl(info_ptr, key, buflen, thread_level_s, &flag);
+    MPIR_ERR_CHECK(mpi_errno);
+
+    /* Set requested thread level value as output */
+    mpi_errno = thread_level_to_int(thread_level_s, threadlevel);
+    MPIR_ERR_CHECK(mpi_errno);
+
+  fn_exit:
+    if (thread_level_s) {
+        MPL_free(thread_level_s);
+    }
+    return mpi_errno;
+  fn_fail:
+    goto fn_exit;
+}
+
+
 int MPIR_Session_init_impl(MPIR_Info * info_ptr, MPIR_Errhandler * errhandler_ptr,
                            MPIR_Session ** p_session_ptr)
 {
     int mpi_errno = MPI_SUCCESS;
     MPIR_Session *session_ptr = NULL;
 
     int provided;
-    /* Let's try ask for MPI_THREAD_MULTIPLE, but it probably still works with
-     * MPI_THREAD_SINGLE since we use MPL_Initlock_lock for cross-session locks.
+    int requested = MPI_THREAD_MULTIPLE;
+
+    /* Get the thread level requested by the user via info object (if any) */
+    mpi_errno = get_thread_level_from_info(info_ptr, &requested);
+    MPIR_ERR_CHECK(mpi_errno);
+
+    /* Remark on MPI_THREAD_SINGLE: Multiple sessions may run in threads
+     * concurrently, so significant work is needed to support per-session MPI_THREAD_SINGLE.
+     * For now, it probably still works with MPI_THREAD_SINGLE since we use MPL_Initlock_lock
+     * for cross-session locks in MPII_Init_thread.
+     *
+     * TODO: support per-session MPI_THREAD_SINGLE and optimize
      */
-    mpi_errno = MPII_Init_thread(NULL, NULL, MPI_THREAD_MULTIPLE, &provided, &session_ptr);
+    mpi_errno = MPII_Init_thread(NULL, NULL, requested, &provided, &session_ptr);
     MPIR_ERR_CHECK(mpi_errno);
 
     session_ptr->thread_level = provided;
```

Comment:
We can store user-requested thread_level here. This is used to feed back to user on what thread level they "promised".  And in the future, we could use this info to optimize.

Reply:
We have to make sure to be compatible with the standard. Users have to be able to obtain the actual provided thread level via MPI_Session_get_info. So storing whatever the user requested here is not a good idea in my point of view.

I've changed the code a bit to  find a compromise. The provided thread level is stored in `thread_level` and the user-requested thread level (if any) is stored in an additional parameter `thread_level_req` so that we can return both to the user via MPI_Session_get_info (in the worst case this tells a user that MPI understood the requested thread level but cannot do any better than the provided one) and keep the requested thread level for future optimizations.

Reply:
I don't think it's wrong if the user request `MPI_THREAD_SINGLE`, and get `MPI_THREAD_SINGLE`, even though the implementations internally will work with `MPI_THREAD_MULTIPLE`.  For rationale: when user requests a lower thread-level, it is promising its *maximum* thread-level and it only matters if implementation can't accommodate it. For example, if user request `MPI_THREAD_SINGLE`, and there is no significance if `provided` is `MPI_THREAD_SINGLE` or `MPI_THREAD_MULTIPLE`. User code will not change in anyway, but implementation will have a confirmed promise that user code will be bound by `MPI_THREAD_SINGLE`.

Reply:
Ok. I made some small modifications. Are they compliant with your explanations?